lib/widgets/compose_box.dart

```diff
@@ -0,0 +1,281 @@
+import 'package:flutter/material.dart';
+import 'package:zulip/widgets/dialog.dart';
+
+import 'app.dart';
+import '../api/route/messages.dart';
+
+enum TopicValidationError { mandatoryButEmpty, tooLong }
+
+class TopicTextEditingController extends TextEditingController {
+  // TODO: subscribe to this value:
+  //   https://zulip.com/help/require-topics
+  final mandatory = true;
+
+  String textNormalized() {
+    String trimmed = text.trim();
+    return trimmed.isEmpty ? kNoTopicTopic : trimmed;
+  }
+
+  List<TopicValidationError> validationErrors() {
+    List<TopicValidationError> result = [];
+    final normalized = textNormalized();
+    if (mandatory && normalized == kNoTopicTopic) {
+      result.add(TopicValidationError.mandatoryButEmpty);
+    }
+    if (normalized.length > kMaxTopicLength) {
+      result.add(TopicValidationError.tooLong);
+    }
+    return result;
+  }
+}
+
+enum ContentValidationError {
+  empty,
+  tooLong,
+  // TODO: upload in progress
+  // TODO: quote-and-reply in progress
+}
+
+class ContentTextEditingController extends TextEditingController {
+  String textNormalized() {
+    return text.trim();
+  }
+
+  List<ContentValidationError> validationErrors() {
+    List<ContentValidationError> result = [];
+    final normalized = textNormalized();
+    if (normalized.isEmpty) {
+      result.add(ContentValidationError.empty);
+    }
+    // normalized.length is actually UTF-16 code units, while the server API
+    // expresses the max in Unicode code points. So this comparison will be
+    // conservative and may cut the user off shorter than necessary.
+    if (normalized.length > kMaxMessageLengthCodePoints) {
+      result.add(ContentValidationError.tooLong);
+    }
+    return result;
+  }
+}
+
+/// The send button for StreamComposeBox.
+class _StreamSendButton extends StatefulWidget {
+  const _StreamSendButton({required this.topicController, required this.contentController});
+
+  final TopicTextEditingController topicController;
+  final ContentTextEditingController contentController;
+
+  @override
+  State<_StreamSendButton> createState() => _StreamSendButtonState();
+}
+
+class _StreamSendButtonState extends State<_StreamSendButton> {
+  late bool topicHasValidationErrors;
+  late bool contentHasValidationErrors;
+
+  topicValueChanged() {
+    setState(() {
+      topicHasValidationErrors = widget.topicController.validationErrors().isNotEmpty;
+    });
+  }
+
+  contentValueChanged() {
+    setState(() {
+      contentHasValidationErrors = widget.contentController.validationErrors().isNotEmpty;
+    });
+  }
+
+  @override
+  void initState() {
+    super.initState();
+
+    topicHasValidationErrors = widget.topicController.validationErrors().isNotEmpty;
+    contentHasValidationErrors = widget.contentController.validationErrors().isNotEmpty;
+
+    widget.topicController.addListener(topicValueChanged);
+    widget.contentController.addListener(contentValueChanged);
```

Comment:
Cool. Yeah, in order to have the send button do its own listening to the text controllers, it needs to be stateful.